File: 128. Longest Consecutive Sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> s;
        for (int num : nums) s.insert(num);
        int maxlen = 0;
        while (!s.empty()) {
            int len = 0;
            int pick = *s.begin();
            int left = pick, right = pick + 1;
            while (s.count(left) > 0) {
                ++len;
                s.erase(left--);
            }
            
            while (s.count(right) > 0) {
                ++len;
                s.erase(right++);
            }
            
            maxlen = max(maxlen, len);
        }
        
        return maxlen;
    }
};

File: 128. Longest Consecutive Sequence.py
# Time: O(n), space: O(n)
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numset = set(nums)
        
        def expand(num, step):
            l = 0
            while num in numset:
                numset.remove(num)
                num += step
                l += 1
            return l
        
        maxlen = 0
        while numset:
            num = next(iter(numset))
            maxlen = max(maxlen, expand(num, -1) + expand(num + 1, 1))
        return maxlen
